### PR TITLE
 merge lyft to cloudanix-megre-upstream

### PR DESCRIPTION
**Pull request by sagarnavale**

### CODE REVIEW

File: cartography/intel/aws/apigateway.py [hunk 1]
```diff
@@ -35,24 +33,16 @@ def get_client_certificates(boto3_session: boto3.session.Session, region: str) -
...
+            certificate['region'] = region
+            certificate['arn'] = f"arn:aws:apigateway:{region}:{current_aws_account_id}:clientcertificates/{certificate['clientCertificateId']}"
+
         return certificates
 
     except ClientError as e:
         logger.error(f'Failed to call Apigateway get_client_certificates: {region} - {e}')
         return certificates
 
-@timeit
```
Comment (ajinkya1986) [inline cartography/intel/aws/apigateway.py:44]: is this function removed or still there

File: cartography/intel/aws/apigateway.py [hunk 2]
```diff
@@ -35,24 +33,16 @@ def get_client_certificates(boto3_session: boto3.session.Session, region: str) -
...
-def transform_client_certificates(certs: List[Dict], region: str, account_id: str) -> List[Dict]:
-    certificates = []
```
Comment (ajinkya1986) [inline cartography/intel/aws/apigateway.py:46]: is this method removed or still there

**Review by ajinkya1986 — COMMENTED** [review]
Make sure the transform functions are not removed. Rest looks good. @ganeshjagdale9545  can you second review it

File: cartography/intel/gcp/iam.py [hunk 3]
```diff
@@ -1,24 +1,106 @@
...
+        err = json.loads(e.content.decode('utf-8'))['error']
+        if err['status'] == 'PERMISSION_DENIED':
+            logger.warning(
+                (
+                    "Could not retrieve groups due to permissions issue. Code: %s, Message: %s"
+                ), err['code'], err['message'],
+            )
+            return []
+        else:
+            raise
```
Comment (ajinkya1986) [inline cartography/intel/gcp/iam.py:103]: These were commented in latest

**Review by ajinkya1986 — COMMENTED** [review]
Please check if old code was removed from cloudanix-all while merging

File: cartography/data/jobs/analysis/aws_foreign_accounts.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+  "statements": [
+    {
+      "__comment": "This analyze AWS accounts we created and tag the ones that are foreign. Foreign accounts are ones that were not in the sync scope",
+      "query": "MATCH (foreign:AWSAccount) where NOT EXISTS(foreign.inscope) SET foreign.foreign = true",
```
Comment (ganeshjagdale9545) [inline cartography/data/jobs/analysis/aws_foreign_accounts.json:5]: convert this query to`CloudanixWorkspace ` specific

File: cartography/data/jobs/analysis/aws_foreign_accounts.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+      "iterative": false
+    },
+    {
+      "__comment": "Remove accounts that were set with foreign and inscope. This can happen as we finish the list of sync accounts through assume role mapping and vpc peering",
+      "query": "MATCH (a:AWSAccount) where EXISTS(a.inscope) AND EXISTS(a.foreign) REMOVE a.foreign",
```
Comment (ganeshjagdale9545) [inline cartography/data/jobs/analysis/aws_foreign_accounts.json:10]: same as above

File: cartography/data/jobs/analysis/aws_lambda_ecr.json [hunk 6]
```diff
@@ -3,7 +3,7 @@
     "statements": [
         {
             "__comment": "Create HAS_IMAGE realtionship from lambda functions to the associated ECR image",
-            "query": "MATCH (l:AWSLambda), (e:ECRImage)\nWHERE e.digest = 'sha256:' + l.codesha256\nMERGE (l)-[:HAS]->(e)",
+            "query":"MATCH (l:AWSLambda) \n WITH COLLECT(l) as lmbda_list \n UNWIND lmbda_list as lmbda \n MATCH (e:ECRImage) \n WHERE e.digest = 'sha256:' + lmbda.codesha256 \n MERGE (lmbda)-[r:HAS]->(e) \n SET r.lastupdated = $UPDATE_TAG",
```
Comment (ganeshjagdale9545) [inline cartography/data/jobs/analysis/aws_lambda_ecr.json:6]: same as above

File: cartography/data/jobs/cleanup/aws_import_inspector_cleanup.json [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        "iterative": true,
+        "iterationsize": 100
+      },
+      {
+        "query": "MATCH (:AWSAccount{id: $AWS_ID})-[:RESOURCE]->(n:AWSInspectorFinding) WHERE n.lastupdated <> $UPDATE_TAG WITH n LIMIT $LIMIT_SIZE DETACH DELETE (n)",
+        "iterative": true,
+        "iterationsize": 100
+      },
+      {
+        "query": "MATCH (:AWSAccount{id: $AWS_ID})-[:RESOURCE]->(n:AWSInspectorPackage) WHERE n.lastupdated <> $UPDATE_TAG WITH n LIMIT $LIMIT_SIZE DETACH DELETE (n)",
```
Comment (ganeshjagdale9545) [inline cartography/data/jobs/cleanup/aws_import_inspector_cleanup.json:29]: convert all above queries  to `CloudanixWorkspace` specific

File: cartography/data/jobs/cleanup/aws_import_ssm_cleanup.json [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+      "iterative": true,
+      "iterationsize": 100
+    },
+    {
+      "query": "MATCH (:AWSAccount{id: $AWS_ID})-[:RESOURCE]->(n:SSMInstancePatch) WHERE n.lastupdated <> $UPDATE_TAG WITH n LIMIT $LIMIT_SIZE DETACH DELETE (n)",
+      "iterative": true,
+      "iterationsize": 100
+    },
+    {
+      "query": "MATCH (:AWSAccount{id: $AWS_ID})-[:RESOURCE]->(:EC2Instance)-[r:HAS_PATCH]->(:SSMInstancePatch) WHERE r.lastupdated <> $UPDATE_TAG WITH r LIMIT $LIMIT_SIZE DELETE (r)",
```
Comment (ganeshjagdale9545) [inline cartography/data/jobs/cleanup/aws_import_ssm_cleanup.json:19]: convert all above queries  to `CloudanixWorkspace` specific

File: cartography/intel/aws/__init__.py [hunk 9]
```diff
@@ -133,6 +135,15 @@ def _sync_one_account(
...
     )
 
+    merge_module_sync_metadata(
+        neo4j_session,
+        group_type='AWSAccount',
+        group_id=current_aws_account_id,
+        synced_type='AWSAccount',
+        update_tag=update_tag,
+        stat_handler=stat_handler,
+    )
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/__init__.py:145]: @mpurusottamc before use this we need read up around this

File: cartography/intel/aws/__init__.py [hunk 10]
```diff
@@ -300,25 +315,27 @@ def start_aws_ingestion(neo4j_session: neo4j.Session, config: Config) -> None:
...
+        config.aws_best_effort_mode,
         requested_syncs,
     )
 
-    run_analysis_job(
-        'aws_ec2_asset_exposure.json',
-        neo4j_session,
-        common_job_parameters,
-    )
+    if sync_successful:
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/__init__.py:327]: to use this logic `return True` from `_sync_multiple_accounts` method

File: cartography/intel/aws/__init__.py [hunk 11]
```diff
@@ -300,25 +315,27 @@ def start_aws_ingestion(neo4j_session: neo4j.Session, config: Config) -> None:
...
-        neo4j_session,
-        common_job_parameters,
-    )
+        run_analysis_job(
+            'aws_ec2_keypair_analysis.json',
+            neo4j_session,
+            common_job_parameters,
+        )
 
     run_analysis_job(
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/__init__.py:340]: also add this in above if statement

File: cartography/intel/aws/ec2/load_balancer_v2s.py [hunk 12]
```diff
@@ -25,7 +25,7 @@ def get_load_balancer_v2_listeners(client: botocore.client.BaseClient, load_bala
             listeners.extend(page['Listeners'])
 
     except Exception as e:
-        logger.warning(f"Failed retrieve load balancer listeners for region - {region}. Error - {e}")
+        logger.warning(f"Failed retrieve load balancer listeners for region - . Error - {e}")
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/ec2/load_balancer_v2s.py:28]: add region

File: cartography/intel/aws/ec2/snapshots.py [hunk 13]
```diff
@@ -5,6 +5,7 @@
 
 import boto3
 import neo4j
+from botocore.exceptions import ClientError
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/ec2/snapshots.py:8]: Imported twice

File: cartography/intel/aws/ec2/snapshots.py [hunk 14]
```diff
@@ -15,9 +16,21 @@
 logger = logging.getLogger(__name__)
 aws_console_link = AWSLinker()
 
+
+@timeit
+def get_snapshots_in_use(neo4j_session: neo4j.Session, region: str, current_aws_account_id: str) -> List[str]:
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/ec2/snapshots.py:21]: this function doesn’t used anywhere

File: cartography/intel/aws/ec2/snapshots.py [hunk 15]
```diff
@@ -15,9 +16,21 @@
...
+    RETURN v.snapshotid as snapshot
+    """
+    results = neo4j_session.run(query, AWS_ACCOUNT_ID=current_aws_account_id, Region=region)
+    return [r['snapshot'] for r in results if r['snapshot']]
+
+
 @timeit
 @aws_handle_regions
-def get_snapshots(boto3_session: boto3.session.Session, region: str) -> List[Dict]:
+def get_snapshots(boto3_session: boto3.session.Session, region: str, in_use_snapshot_ids: List[str]) -> List[Dict]:
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/ec2/snapshots.py:33]: `in_use_snapshot_ids ` is unused parameter

File: cartography/intel/aws/iam.py [hunk 16]
```diff
@@ -997,111 +730,56 @@ def sync_user_access_keys(
...
-# Steps to get access details:
-# 1.generate service last accessed details by ARN (or overall?)
-# 2.
-#
-#
+    parts = arn.split(":")
+    if len(parts) < 4:
+        return ""
+    else:
+        return parts[4]
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/iam.py:785]: Verify this changes with our branch most of the our cloudanix logic removed

File: cartography/intel/aws/lambda_function.py [hunk 17]
```diff
@@ -45,47 +45,43 @@ def load_lambda_functions(
...
+        WITH lambda, lf
+        MATCH (owner:AWSAccount{id: $AWS_ACCOUNT_ID})
+        MERGE (owner)-[r:RESOURCE]->(lambda)
+        ON CREATE SET r.firstseen = timestamp()
+        SET r.lastupdated = $aws_update_tag
+        WITH lambda, lf
+        MATCH (role:AWSPrincipal{arn: lf.Role})
+        MERGE (lambda)-[r:STS_ASSUME_ROLE_ALLOW]->(role)
+        ON CREATE SET r.firstseen = timestamp()
+        SET r.lastupdated = $aws_update_tag
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/lambda_function.py:84]: Merge changes cloudanix logic don’t remove it

File: cartography/intel/aws/s3.py [hunk 18]
```diff
@@ -668,13 +749,6 @@ def load_s3_buckets(neo4j_session: neo4j.Session, data: Dict, current_aws_accoun
             AWS_ACCOUNT_ID=current_aws_account_id,
             aws_update_tag=aws_update_tag,
         )
-    merge_module_sync_metadata(
-        neo4j_session,
-        group_type='AWSAccount',
-        group_id=current_aws_account_id,
-        synced_type='S3Bucket',
-        update_tag=aws_update_tag,
-    )
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/s3.py:677]: revert this

File: cartography/intel/aws/sqs.py [hunk 19]
```diff
@@ -40,13 +41,13 @@ def get_sqs_queue_list(boto3_session: boto3.session.Session, region: str) -> Lis
...
         queue_urls: List[str],
-) -> Dict[str, Any]:
+) -> List[Tuple[str, Any]]:
     """
     Iterates over all SQS queues. Returns a dict with url as key, and attributes as value.
     """
     client = boto3_session.client('sqs')
 
-    queue_attributes: Dict[str, Any] = {}
+    queue_attributes = []
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/sqs.py:50]: Should be dict

File: cartography/intel/azure/tag.py [hunk 20]
```diff
@@ -166,23 +166,22 @@ def get_tags_list(
...
     SET t.lastupdated = $update_tag,
     t.value = tag.value,
+    t.name = tag.name
     t.key = tag.key
     WITH t,tag
-    MATCH (:CloudanixWorkspace{id: $WORKSPACE_ID})-[:OWNER]->
-    (:AzureTenant{id: $AZURE_TENANT_ID})-[:RESOURCE]->
-    (:AzureSubscription{id: $AZURE_SUBSCRIPTION_ID})-[*]->(l)
-    where l.id = tag.resource_id
+    MATCH (l) where l.id = tag.resource_id
```
Comment (ganeshjagdale9545) [inline cartography/intel/azure/tag.py:180]: Don’t remove cloudanix logic

File: cartography/intel/gcp/compute.py [hunk 21]
```diff
@@ -924,11 +924,11 @@ def load_gcp_instances(session: neo4j.Session, instances_list: List[Dict], gcp_u
...
-    # for instance in instances_list:
-    #     _attach_instance_tags(session, instance, gcp_update_tag)
-    #     _attach_gcp_nics(session, instance, gcp_update_tag)
-    #     _attach_gcp_vpc(session, instance['partial_uri'], gcp_update_tag)
-    #     _attach_instance_service_account(session, instance, gcp_update_tag)
+    for instance in instances_list:
+        _attach_instance_tags(session, instance, gcp_update_tag)
+        _attach_gcp_nics(session, instance, gcp_update_tag)
+        _attach_gcp_vpc(session, instance['partial_uri'], gcp_update_tag)
+        _attach_instance_service_account(session, instance, gcp_update_tag)
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/compute.py:931]: revert this

File: cartography/intel/gcp/iam.py [hunk 22]
```diff
@@ -498,611 +387,366 @@ def cleanup_roles(neo4j_session: neo4j.Session, common_job_parameters: Dict) ->
...
-            set_used_state(neo4j_session, project_id, common_job_parameters, gcp_update_tag)
-    else:
-        bindings = get_policy_bindings(crm_v1, crm_v2, project_id)
-        # users_from_bindings, groups_from_bindings, domains_from_bindings = transform_bindings(bindings, project_id)
-        load_bindings(neo4j_session, bindings, project_id, gcp_update_tag)
-        set_used_state(neo4j_session, project_id, common_job_parameters, gcp_update_tag)
-
-    toc = time.perf_counter()
-    logger.info(f"Time to process IAM: {toc - tic:0.4f} seconds")
+    load_bindings(neo4j_session, bindings, project_id, gcp_update_tag)
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/iam.py:752]: Verify this file with our branch

**Review by ganeshjagdale9545 — CHANGES_REQUESTED** [review]
few changes needed

File: cartography/intel/aws/resources.py [hunk 23]
```diff
@@ -59,6 +60,8 @@
     'eks': eks.sync,
     'ec2:launch_templates': sync_ec2_launch_templates,
     'ec2:autoscalinggroup': sync_ec2_auto_scaling_groups,
+    # `ec2:instance` must be included before `ssm` and `ec2:images`,
```
Comment (ganeshjagdale9545) [inline cartography/intel/aws/resources.py:63]: ssm not included here

File: cartography/intel/gcp/iam.py [hunk 24]
```diff
@@ -47,1062 +86,1878 @@ def get_service_accounts(iam: Resource, project_id: str) -> List[Dict]:
...
+        sync_compute_disks(neo4j_session, compute, project_id, zones,
+                           gcp_update_tag, common_job_parameters)
+        sync_gcp_https_proxies(neo4j_session, compute, project_id,
+                               gcp_update_tag, common_job_parameters)
+        sync_gcp_ssl_proxies(neo4j_session, compute, project_id,
+                             gcp_update_tag, common_job_parameters)
 
     toc = time.perf_counter()
-    logger.info(f"Time to process IAM: {toc - tic:0.4f} seconds")
+    logger.info(f"Time to process Compute: {toc - tic:0.4f} seconds")
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/iam.py:1963]: Verify this file most of the cloudanix logic removed

**Review by ganeshjagdale9545 — CHANGES_REQUESTED** [review]
few changes needed 
also update for cloudanix aws/s3 with lyft aws/s3

File: cartography/intel/aws/s3.py [hunk 25]
```diff
@@ -8,78 +8,38 @@
...
-from cloudconsolelink.clouds.aws import AWSLinker
 
+from cartography.stats import get_stats_client
 from cartography.util import merge_module_sync_metadata
 from cartography.util import run_analysis_job
 from cartography.util import run_cleanup_job
 from cartography.util import timeit
 
 logger = logging.getLogger(__name__)
-aws_console_link = AWSLinker()
```
Comment (mpurusottamc) [inline cartography/intel/aws/s3.py:27]: we should not delete this line. console link api calls would fail.

**Review by mpurusottamc — CHANGES_REQUESTED** [review]
can you fix the s3 console link issue? I will review other files in the meantime

**Comment by mpurusottamc** [comment]
@sagarnavale also, resolve the conflicts

**Comment by mpurusottamc** [comment]
merging to fix things on local machine and deploy code to prod.

**Pull request merged**